tool/populate_pub_cache.dart

```diff
@@ -0,0 +1,110 @@
+// Copyright (c) 2021, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:io';
+
+final _client = HttpClient();
+
+/// Fetches and populates a local .pub-cache directory without using `pub`
+/// client and `pub.dev` services.
+Future<void> main(List<String> args) async {
+  final baseDir = File(Platform.script.toFilePath()).parent.parent;
+  print('Base directory: $baseDir');
+
+  final pubCacheDir = Directory('${baseDir.path}/build/pub-cache');
+  print('PUB_CACHE_DIR: ${pubCacheDir.path}');
+  pubCacheDir.createSync(recursive: true);
+
+  final multiVersions = <String, Set<String>>{};
+  final lockFiles = baseDir
+      .listSync(recursive: true)
+      .whereType<File>()
+      .where((f) => !f.path.contains('/pkg/code_coverage/'))
+      .where((f) => !f.path.contains('/pkg/pub_integration/'))
+      .where((f) => f.path.endsWith('pubspec.lock'));
+  for (final f in lockFiles) {
+    print('Parsing ${f.path} ...');
+    final pvs = _parsePubspecLockSync(f);
+    for (final package in pvs.keys) {
+      multiVersions.putIfAbsent(package, () => <String>{}).add(pvs[package]!);
+    }
+  }
+
+  final packages = multiVersions.keys.toList()..sort();
+  for (final package in packages) {
+    final versions = multiVersions[package]!.toList()..sort();
+    print('Found $package: $versions ...');
+    for (final version in versions) {
+      await _downloadInto(package, version, pubCacheDir);
```

Comment:
Maybe wrap this with a rudimentary try / catch and retry...
We don't need exponential backoff, just sleep 3s before trying again.

Reply:
Also in the `catch` block it might be smart to delete the folder we expected this to be written to... since there could be a partially extracted package there..